--- kube-containerd/1.4/config.default.toml     2020-09-01 17:52:14.095499800 +0800
+++ kube-containerd/1.5/config.default.toml     2020-09-22 11:09:54.593385700 +0800
@@ -126,8 +126,14 @@
     platforms = ["linux/amd64"]
   [plugins."io.containerd.service.v1.diff-service"]
     default = ["walking"]
+  [plugins."io.containerd.snapshotter.v1.btrfs"]
+    root_path = ""
   [plugins."io.containerd.snapshotter.v1.devmapper"]
     root_path = ""
     pool_name = ""
     base_image_size = ""
     async_remove = false
+  [plugins."io.containerd.snapshotter.v1.native"]
+    root_path = ""
+  [plugins."io.containerd.snapshotter.v1.overlayfs"]
+    root_path = ""
